lib/src/util/bolt_helpers.cc

```diff
@@ -0,0 +1,195 @@
+#include <pxp-agent/util/bolt_helpers.hpp>
+#include <pxp-agent/configuration.hpp>
+#include <pxp-agent/module.hpp>
+
+#include <boost/algorithm/hex.hpp>
+
+#define LEATHERMAN_LOGGING_NAMESPACE "puppetlabs.pxp_agent.util.bolt_helpers"
+#include <leatherman/logging/logging.hpp>
+
+#include <openssl/evp.h>
+
+namespace fs       = boost::filesystem;
+namespace alg      = boost::algorithm;
+namespace lth_curl = leatherman::curl;
+namespace lth_jc   = leatherman::json_container;
+namespace lth_loc  = leatherman::locale;
+
+namespace PXPAgent {
+namespace Util {
+
+
+    // NIX_DIR_PERMS is defined in pxp-agent/configuration
+    #define NIX_TASK_FILE_PERMS NIX_DIR_PERMS
+
+    // Downloads a file if it does not already exist on the filesystem. A check is made
+    // on the filesystem to determine if the file at destination already exists and if
+    // it already matches the sha256 provided with the file. If the file already exists
+    // the function immediately returns.
+    //
+    // If the file does not exist attempt to download with leatherman.curl. Once the
+    // download finishes a sha256 check occurs to ensure file contents are correct. Then
+    // the file is moved to destination with boost::filesystem::rename.
+    //
+    // If the destination param is empty: the function assumes the file should
+    // remain in the cache_dir and will be renamed to cache_dir/filename
+    fs::path downloadFileFromMaster(const std::vector<std::string>& master_uris,
+                                uint32_t connect_timeout,
+                                uint32_t timeout,
+                                lth_curl::client& client,
+                                const fs::path& cache_dir,
+                                const fs::path& destination,
```

Comment:
IS there a reason that this isn't just called `filepath`? Is there an external API that needs this name?

Reply:
destination can be left empty, in which case filepath defaults to cache_dir/filename.

Reply:
Understood; the question is why do we pass in one name when all we do is assign it to a different one? Is it because we want to keep it a `const`?

Reply:
oh, yeah, now that I think about it you're probably right. There's no need for this to be a const and for the function not to just modify it.

Reply:
I don't like how I wrote this so I've just moved the destination compilation away from this function. Now destination is always where this function downloads to, and the calling function in task defines destination as cache_dir / filename.

That should be more explicit anyway